src/nvim/lua/treesitter.c

```diff
@@ -1322,7 +1320,9 @@ static int node_tree(lua_State *L)
     return 0;
   }
 
-  push_tree(L, (TSTree *)node.tree, false);
+  lua_getfenv(L, 1);  // [udata, reftable]
+  lua_rawgeti(L, -1, 1);  // [udata, reftable, tree_udata]
```

Comment:
It took me a while to understand what's going on here, but after studying this file and `luaref.txt` I think I understand it. Maybe we can introduce some abstraction to make this clearer? Or at least more comments.

I hope this is correct:

Basically, in order for trees to be refcounted, we need to store them in some table in Lua, doesn't matter where. Since environments associated with userdata have no meaning in Lua (see `:h lua-environments`), we can store such a table directly onto the userdata we want to refcount.

Nodes also need this reftable since their lifetime is dependent on the tree they come from. So whenever we push a node into Lua we need to make sure we set its environment to the reftable of the tree it came from.

Reply:
It's really just a reference from a node to the tree which contains it. but lua 5.1 has a design flaw where a userdata cannot contain another userdata as its uservalue, thus a dummy table is needed. this should be mentioned in comment where the table is first constructed..

Reply:
Yes, the point is to say _what_ is done and _why_  , not to provide a detailed historical account for how lua has stumbled the development of uservalues from one version to another.

> For Lua C API + treesitter experts, this isn't complicated.

Not being an expert unfortunately has not stopped anyone from just noting the presence `lua_setfenv()`, completely skip over any surrounding code and any commends and immediately go into lecture mode and offer their very important opinion about "function environments" or nvim using deprecated functions or whatever..

As it happens lua itself considers 5.1 to be ***dead*** ("y u no upgrade to 5.4 already :zany_face: "), and so any improvements or corrections to the 5.1 manual will be consider out of scope for them. Which makes we wonder if we should take the liberty to make improvements to `luaref.txt` ourselves. These could be marked as "amendments" or something if we want to distinguish them from the official text.

Reply:
Raised #24943.